guides/keras_core/migration_guide_tf_keras_3_0.py

```diff
@@ -0,0 +1,877 @@
+# dummy data
+x_test = np.random.uniform(size=[10, 10])
+y_test = np.random.uniform(size=[10, 5])
+multi_output_model = keras.Model(inputs=inputs, outputs=[output_1, output_2])
+multi_output_model.compile(
+    optimizer="adam",
+    loss="categorical_crossentropy",
+    metrics=["categorical_crossentropy", "categorical_crossentropy"],
+)
+multi_output_model.evaluate(x_test, y_test)
+
+
+"""
+### Variables tracking
+
+The following snippet of code will show that the tf.Variables are not being tracked.
+
+```python
+class MyCustomLayer(keras.layers.Layer):
+    def __init__(self, units):
+        super().__init__()
+        self.units = units
+
+    def build(self, input_shape):
+        input_dim = input_shape[-1]
+        self.w = tf.Variable(initial_value=tf.zeros([input_dim, self.units]))
+        self.b = tf.Variable(initial_value=tf.zeros([self.units,]))
+
+    def call(self, inputs):
+        return keras.ops.matmul(inputs, self.w) + self.b
+
+
+layer = MyCustomLayer(3)
+data = np.random.uniform(size=[3, 3])
+model = keras.models.Sequential([layer])
+model.compile(optimizer="adam", loss="mse")
+model.fit(data, data)
+# The model does not have any trainable variables
+for layer in model.layers:
+    print(layer.trainable_variables)
+```
+
+`UserWarning: The model does not have any trainable weights.
+  warnings.warn("The model does not have any trainable weights.")`
+"""
+
+"""
+Here is how you fix it:
+
+use `self.add_weight()` method or use a `keras.Variable` instead.
+
+"""
+
+
+class MyCustomLayer(keras.layers.Layer):
+    def __init__(self, units):
+        super().__init__()
+        self.units = units
+
+    def build(self, input_shape):
+        input_dim = input_shape[-1]
+        self.w = self.add_weight(
+            shape=[input_dim, self.units],
+            initializer="zeros",
+        )
+        self.b = self.add_weight(
+            shape=[
+                self.units,
+            ],
+            initializer="zeros",
+        )
+
+    def call(self, inputs):
+        return keras.ops.matmul(inputs, self.w) + self.b
+
+
+layer = MyCustomLayer(3)
+data = np.random.uniform(size=[3, 3])
+model = keras.models.Sequential([layer])
+model.compile(optimizer="adam", loss="mse")
+model.fit(data, data)
+# Verify that the variables are now being tracked
+for layer in model.layers:
+    print(layer.trainable_variables)
+
+"""
+### None entries are not allowed as part of nested (e.g. list/tuples) tensor arguments in
+Layer.call(), nor as part of call() return values.
+
+The following snippet of code will reproduce the error.
+
+```python
+class CustomLayer(keras.layers.Layer):
+    def __init__(self):
+        super().__init__()
+
+    def call(self, inputs):
+      return  inputs["foo"]
+
+layer = CustomLayer()
+inputs = {
+    "foo": keras.Input(shape=(1,), name="foo"),
+    "bar": {
+        "baz": None,
+    },
+}
+layer(inputs)
+```
+"""
+
+"""
+Here is how you will fix it:
+
+Replace `None` with a value or Keras Tensor
```

Comment:
This answer is incomplete -- the None was there for a reason. If the argument is optional then structure it as a separate argument, e.g. `call(self, x1, x2=None)`

Reply:
Added more context and the proposed solution.